File: GPU-sizer/v2/js/calculator.js
// Performance calculator for LLM sizing on Nutanix Enterprise AI.
//
// All memory math is done consistently in GiB (2^30 bytes). Compute/latency
// math accounts for:
//   - MoE models: compute & decode bandwidth use ACTIVE params/token, while
//     capacity (weights in memory) uses TOTAL params.
//   - Precision: decode (memory-bound) scales with the weight byte width, so
//     INT8/INT4 quantization speeds up token generation.
//   - Real-world efficiency: theoretical peak FLOPS/bandwidth are derated by
//     tunable utilization factors (MFU / MBU).

const BYTES_IN_GiB = 1_073_741_824;

// Default real-world utilization factors (override via the calculator options).
const DEFAULT_COMPUTE_EFFICIENCY = 0.30;   // MFU: fraction of peak FLOPS reached during prefill.
const DEFAULT_BANDWIDTH_EFFICIENCY = 0.70; // MBU: fraction of peak mem bandwidth reached during decode.
const DEFAULT_MEMORY_OVERHEAD = 1.15;      // Activations + CUDA/runtime overhead on top of weights.

class PerformanceMetrics {
    constructor(kv_cache_tokens, prefill_time_per_token_ms, tpot_ms, ttft_s,
                e2e_latency_s, throughput_tokens_per_s, system_throughput_tokens_per_s = "N/A") {
        this.kv_cache_tokens = kv_cache_tokens;
        this.prefill_time_per_token_ms = prefill_time_per_token_ms;
        this.tpot_ms = tpot_ms;
        this.ttft_s = ttft_s;
        this.e2e_latency_s = e2e_latency_s;
        this.throughput_tokens_per_s = throughput_tokens_per_s;
        this.system_throughput_tokens_per_s = system_throughput_tokens_per_s;
    }
}

class PerformanceCalculator {
    /**
     * @param {number} num_gpu
     * @param {PrecisionSpec} precision
     * @param {object} [options] { computeEfficiency, bandwidthEfficiency, memoryOverhead }
     */
    constructor(num_gpu, precision, options = {}) {
        this.num_gpu = Math.max(1, num_gpu | 0);
        this.precision = precision;
        this.computeEfficiency = options.computeEfficiency ?? DEFAULT_COMPUTE_EFFICIENCY;
        this.bandwidthEfficiency = options.bandwidthEfficiency ?? DEFAULT_BANDWIDTH_EFFICIENCY;
        this.memoryOverhead = options.memoryOverhead ?? DEFAULT_MEMORY_OVERHEAD;
    }

    // ---------- Helpers ----------
    activeParams(model) {
        // Fall back to total params for dense models / older specs.
        return (model.active_params_billion && model.active_params_billion > 0)
            ? model.active_params_billion
            : model.params_billion;
    }

    // ---------- Memory / capacity (all GiB) ----------
    kvCacheSizePerTokenGib(model) {
        if (!model.n_heads || !model.n_layers) return 0;
        const d_head = model.d_model / model.n_heads;
        // 2x for K and V; GQA/MQA uses n_kv_heads (< n_heads).
        return (2 * model.n_layers * model.n_kv_heads * d_head * this.precision.kv_bytes) / BYTES_IN_GiB;
    }

    weightsMemoryGib(model) {
        // Total (not active) params live in memory.
        return (model.params_billion * 1e9 * this.precision.weight_bytes) / BYTES_IN_GiB;
    }

    // Weights + framework/activation overhead reserved before any KV cache.
    reservedMemoryGib(model) {
        return this.weightsMemoryGib(model) * this.memoryOverhead;
    }

    totalMemoryFootprintGib(model, n_concurrent, context_window) {
        const kv_total = this.kvCacheSizePerTokenGib(model) * context_window * n_concurrent;
        return this.reservedMemoryGib(model) + kv_total;
    }

    totalGpuMemoryGib(gpu) {
        // GPU memory is quoted in GB (10^9); convert to GiB for consistent math.
        return (this.num_gpu * gpu.memory_gb * 1e9) / BYTES_IN_GiB;
    }

    maxKvTokens(gpu, model) {
        const kv_per_token = this.kvCacheSizePerTokenGib(model);
        if (kv_per_token <= 0) return 0;
        const usable = this.totalGpuMemoryGib(gpu) - this.reservedMemoryGib(model);
        return Math.floor(Math.max(0.0, usable / kv_per_token));
    }

    fitsMemory(gpu, model, n_concurrent, context_window) {
        return this.totalMemoryFootprintGib(model, n_concurrent, context_window)
            <= this.totalGpuMemoryGib(gpu);
    }

    // ---------- Time / performance ----------
    // Prefill is compute-bound: ~2 FLOPs per active parameter per token.
    prefillTimePerTokenMs(model, gpu) {
        const flops_per_token = 2 * this.activeParams(model) * 1e9;
        const peak_flops = gpu.fp16_tflops * 1e12 * this.computeEfficiency * this.num_gpu;
        if (peak_flops <= 0) return Infinity;
        return (flops_per_token / peak_flops) * 1000; // seconds -> ms
    }

    // Decode is memory-bandwidth-bound: read active weights (at weight precision) per token.
    tpotMs(model, gpu) {
        const bytes_per_token = this.activeParams(model) * 1e9 * this.precision.weight_bytes;
        const peak_bw = gpu.memory_bandwidth_gbps * 1e9 * this.bandwidthEfficiency * this.num_gpu;
        if (peak_bw <= 0) return Infinity;
        return (bytes_per_token / peak_bw) * 1000; // seconds -> ms
    }

    ttftS(prefill_ms, tpot_ms, prompt_tokens) {
        // Prefill the whole prompt, then emit the first token.
        return (prompt_tokens * prefill_ms + tpot_ms) / 1000.0;
    }

    e2eLatencyS(prefill_ms, tpot_ms, prompt_tokens, response_tokens) {
        return (prompt_tokens * prefill_ms + response_tokens * tpot_ms) / 1000.0;
    }

    computeMetrics(model, gpu, prompt_tokens, response_tokens, n_concurrent = 1) {
        const kv_tokens = this.maxKvTokens(gpu, model);
        const prefill_ms = this.prefillTimePerTokenMs(model, gpu);
        const tpot_ms = this.tpotMs(model, gpu);

        if (!isFinite(prefill_ms) || !isFinite(tpot_ms) || prefill_ms < 0 || tpot_ms < 0) {
            return new PerformanceMetrics(kv_tokens, "OOM", "OOM", "OOM", "OOM", "OOM", "OOM");
        }

        const ttft_s = this.ttftS(prefill_ms, tpot_ms, prompt_tokens);
        const e2e_s = this.e2eLatencyS(prefill_ms, tpot_ms, prompt_tokens, response_tokens);

        // Per-request output throughput (single stream).
        const per_request = e2e_s > 0 ? response_tokens / e2e_s : "OOM";
        // System throughput: batched decode amortizes weight reads across
        // concurrent requests, so aggregate decode rate ~ concurrency / tpot.
        const per_user_decode = tpot_ms > 0 ? 1000 / tpot_ms : 0;
        const system = per_user_decode > 0 ? per_user_decode * Math.max(1, n_concurrent) : "OOM";

        return new PerformanceMetrics(
            kv_tokens, prefill_ms, tpot_ms, ttft_s, e2e_s, per_request, system
        );
    }
}

// Export for use in other modules (Node/testing).
if (typeof module !== 'undefined' && module.exports) {
    module.exports = {
        PerformanceCalculator,
        PerformanceMetrics,
        BYTES_IN_GiB,
        DEFAULT_COMPUTE_EFFICIENCY,
        DEFAULT_BANDWIDTH_EFFICIENCY,
        DEFAULT_MEMORY_OVERHEAD,
    };
}

File: GPU-sizer/v2/better_llm_sizer/calculator.py
from dataclasses import dataclass
from typing import Union

from .specs import ModelSpec, GPUSpec, PrecisionSpec


BYTES_IN_GiB = 1_073_741_824


@dataclass
class PerformanceMetrics:
    kv_cache_tokens: int
    prefill_time_per_token_ms: Union[float, str]
    tpot_ms: Union[float, str]
    ttft_s: Union[float, str]
    e2e_latency_s: Union[float, str]
    throughput_tokens_per_s: Union[float, str]


class PerformanceCalculator:
    def __init__(self, num_gpu: int, precision: PrecisionSpec):
        self.num_gpu = num_gpu
        self.precision = precision

    # ---------- Memory / capacity ----------
    def kv_cache_size_per_token_gib(self, model: ModelSpec) -> float:
        # 2x for K and V
        d_head = model.d_model / model.n_heads
        return (
            2 * model.n_layers * model.n_kv_heads * d_head * self.precision.kv_bytes
        ) / BYTES_IN_GiB

    def weights_memory_gb(self, model: ModelSpec) -> float:
        # params_billion * 1e9 params * bytes -> GiB approximated as GB for simplicity
        # original repo used: params_billion * 2 (GB) for FP16. Keep same unit basis (GB)
        bytes_per_billion_params_gb = self.precision.weight_bytes
        return model.params_billion * bytes_per_billion_params_gb

    def total_memory_footprint_gb(
        self, model: ModelSpec, n_concurrent: int, context_window: int
    ) -> float:
        kv_per_token_gib = self.kv_cache_size_per_token_gib(model)
        kv_total_gib = kv_per_token_gib * context_window * n_concurrent
        # Convert GiB to GB roughly 1:1 for this estimator like original (keeps consistency)
        return kv_total_gib + self.weights_memory_gb(model)

    def max_kv_tokens(self, gpu: GPUSpec, model: ModelSpec) -> int:
        kv_per_token_gib = self.kv_cache_size_per_token_gib(model)
        if kv_per_token_gib <= 0:
            return 0
        total_mem_gb = self.num_gpu * gpu.memory_gb
        # Convert GB to GiB roughly 1:1 to maintain simplicity
        usable_gb = max(0.0, total_mem_gb - self.weights_memory_gb(model))
        return int(max(0.0, usable_gb / kv_per_token_gib))

    def fits_memory(
        self, gpu: GPUSpec, model: ModelSpec, n_concurrent: int, context_window: int
    ) -> bool:
        return self.total_memory_footprint_gb(model, n_concurrent, context_window) <= (
            self.num_gpu * gpu.memory_gb
        )

    # ---------- Time / performance ----------
    def prefill_time_per_token_ms(self, model: ModelSpec, gpu: GPUSpec) -> float:
        # Same scaling as original: 2 * params_billion / num_gpu / tflops (ms)
        return (2 * model.params_billion / max(1, self.num_gpu)) / gpu.fp16_tflops

    def tpot_ms(self, model: ModelSpec, gpu: GPUSpec) -> float:
        # Same form as original but clarified units
        return (
            (2 * model.params_billion / max(1, self.num_gpu))
            / max(1e-9, gpu.memory_bandwidth_gbps)
            * 1000
        )

    def e2e_latency_s(
        self, prefill_ms: float, tpot_ms: float, prompt_tokens: int, response_tokens: int
    ) -> float:
        return (prompt_tokens * prefill_ms + response_tokens * tpot_ms) / 1000.0

    def ttft_s(self, prefill_ms: float, tpot_ms: float, prompt_tokens: int) -> float:
        # Correct TTFT: prefill of entire prompt + one token generation
        return (prompt_tokens * prefill_ms + tpot_ms) / 1000.0

    def compute_metrics(
        self, model: ModelSpec, gpu: GPUSpec, prompt_tokens: int, response_tokens: int
    ) -> PerformanceMetrics:
        kv_tokens = self.max_kv_tokens(gpu, model)
        prefill_ms = self.prefill_time_per_token_ms(model, gpu)
        tpot_ms = self.tpot_ms(model, gpu)

        if prefill_ms < 0 or tpot_ms < 0:
            return PerformanceMetrics(
                kv_cache_tokens=kv_tokens,
                prefill_time_per_token_ms="OOM",
                tpot_ms="OOM",
                ttft_s="OOM",
                e2e_latency_s="OOM",
                throughput_tokens_per_s="OOM",
            )

        ttft_s = self.ttft_s(prefill_ms, tpot_ms, prompt_tokens)
        e2e_s = self.e2e_latency_s(prefill_ms, tpot_ms, prompt_tokens, response_tokens)
        thr = response_tokens / e2e_s if e2e_s > 0 else "OOM"

        return PerformanceMetrics(
            kv_cache_tokens=kv_tokens,
            prefill_time_per_token_ms=prefill_ms,
            tpot_ms=tpot_ms,
            ttft_s=ttft_s,
            e2e_latency_s=e2e_s,
            throughput_tokens_per_s=thr,
        )

